8283335 : Add exists and readAttributesIfExists methods to FileSystemProvider

LanceAndersen opened the pull request:
Hi,

Please review the following patch which will:

- Enhance the java.nio.file.spi.FileSystemProvider abstract class to include the methods

  - public boolean exists(Path path, LinkOption... options)
  - public <A extends BasicFileAttributes> A readAttributesIfExists(Path path, Class<A> type, LinkOption... options)


This change allows for providers to provide optimizations when the file's attributes are not needed.

Mach5 tiers 1 - 3  run clean with this change

The CSR may be viewed at [JDK-8283336](https://bugs.openjdk.org/browse/JDK-8283336)


Best,
Lance

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8283335](https://bugs.openjdk.org/browse/JDK-8283335): Add exists and readAttributesIfExists methods to FileSystemProvider
 * [JDK-8283336](https://bugs.openjdk.org/browse/JDK-8283336): Add exists and readAttributesIfExists methods to  FileSystemProvider (**CSR**)


### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9249/head:pull/9249` \
`$ git checkout pull/9249`

Update a local copy of the PR: \
`$ git checkout pull/9249` \
`$ git pull https://git.openjdk.org/jdk pull/9249/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9249`

View PR using the GUI difftool: \
`$ git pr show -t 9249`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9249.diff">https://git.openjdk.org/jdk/pull/9249.diff</a>

</details>

bplb commented on src/java.base/share/classes/java/nio/file/spi/FileSystemProvider.java:
```diff
@@ -1170,4 +1170,122 @@ public abstract Map<String,Object> readAttributes(Path path, String attributes,
+     * are followed. If the option {@link LinkOption#NOFOLLOW_LINKS NOFOLLOW_LINKS}
+     * is present then symbolic links are not followed and the method
+     *  {@link #readAttributes(Path, Class, LinkOption...)} is called
+     *  to determine whether a file exists.
```
Indentation is off at lines 1183-1184.

bplb commented on src/java.base/share/classes/java/nio/file/spi/FileSystemProvider.java:
```diff
@@ -1170,4 +1170,122 @@ public abstract Map<String,Object> readAttributes(Path path, String attributes,
+    }
+
+    /**
+     * Reads a file's attributes as a bulk operation if it exists.
```
Maybe instead "it" put "the file"?

LanceAndersen replied:
I can if you thank that is better.  I followed the wording for `FileSystemProvider::deleteIfExists` trying to be consistent

bplb replied:
Then leave is as is.

bplb commented on src/java.base/unix/classes/sun/nio/fs/UnixFileAttributes.java:
```diff
@@ -74,6 +74,19 @@ static UnixFileAttributes get(UnixPath path, boolean followLinks)
+    // get the UnixFileAttributes for a given file. Returns null if the file does not exist.
+    static UnixFileAttributes getIfExists(UnixPath path) throws UnixException {
+        UnixFileAttributes attrs = new UnixFileAttributes();
+        int errno = UnixNativeDispatcher.stat2(path, attrs);
```
I think that this is the only use of `stat2()`. It could be deleted if this method were replaced with this:

```
    static UnixFileAttributes getIfExists(UnixPath path) throws UnixException {
        UnixFileAttributes attrs = new UnixFileAttributes();
        try {
            UnixNativeDispatcher.stat(path, attrs);
            return attrs;
        } catch (UnixException e) {
            if (e.errno() == UnixConstants.ENOENT) {
                return null;
            }
            throw e;
        }
    }
```

AlanBateman replied:
> I think that this is the only use of `stat2()`. It could be deleted if this method were replaced with this:

The purpose of these additions to the SPI is to improve the performance of user facing methods that don't throw an exception when the file doesn't exist. It needs a stat/equivalent that doesn't throw so this is the reason for stat2. The equivalent on Windows will need to added too.

bplb replied:
Understood.

bplb commented on src/java.base/share/classes/java/nio/file/spi/FileSystemProvider.java:
```diff
@@ -1170,4 +1170,122 @@ public abstract Map<String,Object> readAttributes(Path path, String attributes,
+     * <p> It is implementation specific if all file attributes are read as an
+     * atomic operation with respect to other file system operations.
+     *
+     *  @implSpec
```
One more bad indentation.

ChrisHegarty commented on src/java.base/share/classes/java/nio/file/spi/FileSystemProvider.java:
```diff
@@ -1170,4 +1170,122 @@ public abstract Map<String,Object> readAttributes(Path path, String attributes,
+     * @since 20
+     */
+    public boolean exists(Path path, LinkOption... options) {
+        try {
```
Overall, I think that this is a great change (avoiding the need for various parts of the system to communicate through exceptions).

Trivially, and easily, I often find it useful to add mock-like tests for these specified implementations (implSpec), i.e. just subclass FSP mocking out the abstract methods to ensure that the default implementations invoke the appropriate ones.

LanceAndersen replied:
Thank you for the suggestion Chris

AlanBateman commented:
This should be performance neutral for the default provider and positive for the zip provider. Do we have any micros that could be added?

AlanBateman commented on src/java.base/unix/classes/sun/nio/fs/UnixFileSystemProvider.java:
```diff
@@ -281,10 +300,10 @@ public void checkAccess(Path obj, AccessMode... modes) throws IOException {
+                    case READ -> r = true;
+                    case WRITE -> w = true;
+                    case EXECUTE -> x = true;
+                    default -> throw new AssertionError("Should not get here");
```
The switch should be exhaustive so I assume the default case is not required here.

LanceAndersen replied:
Addressed in next push

AlanBateman commented:
The end date in the copyright header will need updated a few files, I assume you'll do that before this change is integrated.

AlanBateman commented on src/jdk.zipfs/share/classes/jdk/nio/zipfs/ZipFileSystemProvider.java:
```diff
@@ -284,6 +293,15 @@ public OutputStream newOutputStream(Path path, OpenOption... options)
+                                                                    LinkOption... options)
+        throws IOException
+    {
+        return exists(path) ? readAttributes(path, type, options) : null;
```
If you add the following to ZipPath then it would allow you to implement it in one step rather than two:

```
ZipFileAttributes readAttributesIfExits() throws IOException {
    return zfs.getFileAttributes(getResolvedPath());
 }
```

AlanBateman commented:
The implementation changes in latest version address my previous points, thanks!  I don't the test as it duplicates most of the existing of the TestProvider in this directory and the naming is inconsistent with the existing tests. I expect you should be able to just extent TestProvider and override the two methods to record that they can been called.

LanceAndersen commented:
> The implementation changes in latest version address my previous points, thanks! I don't the test as it duplicates most of the existing of the TestProvider in this directory and the naming is inconsistent with the existing tests. I expect you should be able to just extent TestProvider and override the two methods to record that they can been called.

I thought about using TestProvider and then talked myself out of it.  So yes, I can do that.  If you have a preferred name other than `TestOverRideFSPMethods` please let me know

AlanBateman commented on test/jdk/java/nio/file/spi/TestDelegation.java:
```diff
@@ -0,0 +1,152 @@
+        assertEquals(Files.isRegularFile(p), isFile);
+        assertEquals(1, PROVIDER.findCall("exists").size());
+        assertEquals(2, PROVIDER.findCall("readAttributesIfExists").size());
+        PROVIDER.resetCalls();
```
One suggestion is to split these into testExists, testIsDirectory and testIsRegularFile so that the delegation of each method is tested in isolation.

AlanBateman commented:
The updated TestDelegation test is looking a bit better now but I think it would be simplified a lot more by getting rid of the data providers, just aren't needed in this test.

AlanBateman commented:
> Unless you feel this is a must, I would prefer to keep the DataProviders. The benefit I see is the test code does not need to be duplicated per parameter, each test scenario can be run as an individual test so that you do not need extra plumbing to run each test scenario in the unlikely event of a failure.

Okay, but there are a few other things to mention:

One issue is the reset method is called at the end of each test. I think it needs to be at the beginning of the method, moved to a finally block of a try-finally, or maybe @BeforeMethod to reset before each test. The reason is that one test failing will cause the tests that follow to fail too.

The fields aren't constants so looks a bit strange (to me anyway) to be in uppercase. If you rename them then I think the tests would be a bit more readable.

LanceAndersen commented:
> > Unless you feel this is a must, I would prefer to keep the DataProviders. The benefit I see is the test code does not need to be duplicated per parameter, each test scenario can be run as an individual test so that you do not need extra plumbing to run each test scenario in the unlikely event of a failure.
> 
> Okay, but there are a few other things to mention:
> 
> One issue is the reset method is called at the end of each test. I think it needs to be at the beginning of the method, moved to a finally block of a try-finally, or maybe @BeforeMethod to reset before each test. The reason is that one test failing will cause the tests that follow to fail too.

Good catch, added `@BeforeMethod`
> 
> The fields aren't constants so looks a bit strange (to me anyway) to be in uppercase. If you rename them then I think the tests would be a bit more readable.

Ah,  sometimes you see what you want.  I must have had fields on my mind when I did that as I completely agree with you.  

 Both of the above are addressed in the latest update to the PR.

Thank you for the feedback.

AlanBateman reviewed (approved):
I think all my comments have been addressed so I think this is good to go.

LanceAndersen commented:
/integrate
